When `AZURE_CONFIG_DIR` is set, login context gets cleared by setting `random_config_dir=True`

ADO's Azure CLI task contains this step:

```
Setting AZURE_CONFIG_DIR env variable to: /mnt/vss/_work/_temp/.azclitask
```

In Knack, the config dir specified by env var takes higher precedence:

https://github.com/microsoft/knack/blob/e0c14114aea5e4416c70a77623e403773aba73a8/knack/config.py#L44

```py
        self.config_dir = os.environ.get('{}CONFIG_DIR'.format(env_var_prefix), default_config_dir)
```

So when `AZURE_CONFIG_DIR` is set, setting `config_dir` becomes a no-op for `AzCli`:

https://github.com/Azure/azure-cli/blob/c087891017008dd22c49d14742dc7d6480212cfd/src/azure-cli-core/azure/cli/core/mock.py#L44-L46

`self.config.config_dir` in this case is the value of `AZURE_CONFIG_DIR` (`/mnt/vss/_work/_temp/.azclitask`).

During `tearDown`, directory specified by `AZURE_CONFIG_DIR` gets deleted:

https://github.com/Azure/azure-cli/blob/b223dd9d4379d5c633a9f4dba4941ae791531126/src/azure-cli-testsdk/azure/cli/testsdk/base.py#L139-L144

This causes failure in the following `azdev test --live` command:

```
self = <azure.cli.core._profile.Profile object at 0x7f3d731794b0>
subscription = None

    def get_subscription(self, subscription=None):  # take id or name
        subscriptions = self.load_cached_subscriptions()
        if not subscriptions:
>           raise CLIError(_AZ_LOGIN_MESSAGE)
E           knack.util.CLIError: Please run 'az login' to setup account.

src/azure-cli-core/azure/cli/core/_profile.py:546: CLIError
```

Introduced by #25689

--- a/src/azure-cli-core/azure/cli/core/mock.py
+++ b/src/azure-cli-core/azure/cli/core/mock.py
@@ -10,6 +10,7 @@ class DummyCli(AzCli):
     """A dummy CLI instance can be used to facilitate automation"""
     def __init__(self, commands_loader_cls=None, random_config_dir=False, **kwargs):
         import os
+        from unittest.mock import patch
 
         from azure.cli.core import MainCommandsLoader
         from azure.cli.core.commands import AzCliCommandInvoker
@@ -41,16 +42,20 @@ class DummyCli(AzCli):
         else:
             config_dir = GLOBAL_CONFIG_DIR
 
-        super(DummyCli, self).__init__(
-            cli_name='az',
-            config_dir=config_dir,
-            config_env_var_prefix=ENV_VAR_PREFIX,
-            commands_loader_cls=commands_loader_cls or MainCommandsLoader,
-            parser_cls=AzCliCommandParser,
-            logging_cls=AzCliLogging,
-            output_cls=AzOutputProducer,
-            help_cls=AzCliHelp,
-            invocation_cls=AzCliCommandInvoker)
+        # Knack prioritizes the AZURE_CONFIG_DIR env over the config_dir param, we need to set the env var temporarily
+        # to make sure it uses the config_dir param
+        patch_env = {'AZURE_CONFIG_DIR': config_dir} if random_config_dir else {}
+        with patch.dict('os.environ', patch_env):
+            super(DummyCli, self).__init__(
+                cli_name='az',
+                config_dir=config_dir,
+                config_env_var_prefix=ENV_VAR_PREFIX,
+                commands_loader_cls=commands_loader_cls or MainCommandsLoader,
+                parser_cls=AzCliCommandParser,
+                logging_cls=AzCliLogging,
+                output_cls=AzOutputProducer,
+                help_cls=AzCliHelp,
+                invocation_cls=AzCliCommandInvoker)
 
         self.data['headers'] = {}  # the x-ms-client-request-id is generated before a command is to execute
         self.data['command'] = 'unknown'

--- a/src/azure-cli-core/azure/cli/core/mock.py
+++ b/src/azure-cli-core/azure/cli/core/mock.py
@@ -10,6 +10,7 @@ class DummyCli(AzCli):
     """A dummy CLI instance can be used to facilitate automation"""
     def __init__(self, commands_loader_cls=None, random_config_dir=False, **kwargs):
         import os
+        from unittest.mock import patch
 
         from azure.cli.core import MainCommandsLoader
         from azure.cli.core.commands import AzCliCommandInvoker
@@ -26,9 +27,10 @@ class DummyCli(AzCli):
 
         if random_config_dir:
             config_dir = os.path.join(GLOBAL_CONFIG_DIR, 'dummy_cli_config_dir', random_string())
-            # Knack prioritizes the AZURE_CONFIG_DIR env over the config_dir param, and other function may call
+            # Knack prioritizes the AZURE_CONFIG_DIR env over the config_dir param, and other functions may call
             # get_config_dir directly. We need to set the env to make sure the config_dir is used.
-            os.environ['AZURE_CONFIG_DIR'] = config_dir
+            self.env_patch = patch.dict('os.environ', {'AZURE_CONFIG_DIR': config_dir})
+            self.env_patch.start()
 
             # In recording mode, copy login credentials from global config dir to the dummy config dir
             if os.getenv(ENV_VAR_TEST_LIVE, '').lower() == 'true':

--- a/src/azure-cli-core/azure/cli/core/mock.py
+++ b/src/azure-cli-core/azure/cli/core/mock.py
@@ -43,12 +43,10 @@ class DummyCli(AzCli):
                             shutil.copy(os.path.join(GLOBAL_CONFIG_DIR, file), config_dir)
                         except FileNotFoundError:
                             pass
-        else:
-            config_dir = GLOBAL_CONFIG_DIR
 
         super(DummyCli, self).__init__(
             cli_name='az',
-            config_dir=config_dir,
+            config_dir=GLOBAL_CONFIG_DIR,
             config_env_var_prefix=ENV_VAR_PREFIX,
             commands_loader_cls=commands_loader_cls or MainCommandsLoader,
             parser_cls=AzCliCommandParser,

--- a/src/azure-cli-core/azure/cli/core/mock.py
+++ b/src/azure-cli-core/azure/cli/core/mock.py
@@ -10,7 +10,6 @@ class DummyCli(AzCli):
     """A dummy CLI instance can be used to facilitate automation"""
     def __init__(self, commands_loader_cls=None, random_config_dir=False, **kwargs):
         import os
-        from unittest.mock import patch
 
         from azure.cli.core import MainCommandsLoader
         from azure.cli.core.commands import AzCliCommandInvoker
@@ -27,6 +26,9 @@ class DummyCli(AzCli):
 
         if random_config_dir:
             config_dir = os.path.join(GLOBAL_CONFIG_DIR, 'dummy_cli_config_dir', random_string())
+            # Knack prioritizes the AZURE_CONFIG_DIR env over the config_dir param, and other function may call
+            # get_config_dir directly. We need to set the env to make sure the config_dir is used.
+            os.environ['AZURE_CONFIG_DIR'] = config_dir
 
             # In recording mode, copy login credentials from global config dir to the dummy config dir
             if os.getenv(ENV_VAR_TEST_LIVE, '').lower() == 'true':
@@ -42,20 +44,16 @@ class DummyCli(AzCli):
         else:
             config_dir = GLOBAL_CONFIG_DIR
 
-        # Knack prioritizes the AZURE_CONFIG_DIR env over the config_dir param, we need to set the env var temporarily
-        # to make sure it uses the config_dir param
-        patch_env = {'AZURE_CONFIG_DIR': config_dir} if random_config_dir else {}
-        with patch.dict('os.environ', patch_env):
-            super(DummyCli, self).__init__(
-                cli_name='az',
-                config_dir=config_dir,
-                config_env_var_prefix=ENV_VAR_PREFIX,
-                commands_loader_cls=commands_loader_cls or MainCommandsLoader,
-                parser_cls=AzCliCommandParser,
-                logging_cls=AzCliLogging,
-                output_cls=AzOutputProducer,
-                help_cls=AzCliHelp,
-                invocation_cls=AzCliCommandInvoker)
+        super(DummyCli, self).__init__(
+            cli_name='az',
+            config_dir=config_dir,
+            config_env_var_prefix=ENV_VAR_PREFIX,
+            commands_loader_cls=commands_loader_cls or MainCommandsLoader,
+            parser_cls=AzCliCommandParser,
+            logging_cls=AzCliLogging,
+            output_cls=AzOutputProducer,
+            help_cls=AzCliHelp,
+            invocation_cls=AzCliCommandInvoker)
 
         self.data['headers'] = {}  # the x-ms-client-request-id is generated before a command is to execute
         self.data['command'] = 'unknown'

--- a/src/azure-cli-core/azure/cli/core/mock.py
+++ b/src/azure-cli-core/azure/cli/core/mock.py
@@ -29,7 +29,7 @@ class DummyCli(AzCli):
             config_dir = os.path.join(GLOBAL_CONFIG_DIR, 'dummy_cli_config_dir', random_string())
             # Knack prioritizes the AZURE_CONFIG_DIR env over the config_dir param, and other functions may call
             # get_config_dir directly. We need to set the env to make sure the config_dir is used.
-            self.env_patch = patch.dict('os.environ', {'AZURE_CONFIG_DIR': config_dir})
+            self.env_patch = patch.dict(os.environ, {'AZURE_CONFIG_DIR': config_dir})
             self.env_patch.start()
 
             # In recording mode, copy login credentials from global config dir to the dummy config dir

--- a/src/azure-cli-testsdk/azure/cli/testsdk/base.py
+++ b/src/azure-cli-testsdk/azure/cli/testsdk/base.py
@@ -142,6 +142,7 @@ class ScenarioTest(ReplayableTest, CheckerMixin, unittest.TestCase):
         if self.random_config_dir:
             from azure.cli.core.util import rmtree_with_retry
             rmtree_with_retry(self.cli_ctx.config.config_dir)
+            self.cli_ctx.env_patch.stop()
         super(ScenarioTest, self).tearDown()
 
     def create_random_name(self, prefix, length):

